lib/AST/ASTScopeLookup.cpp

```diff
@@ -514,12 +514,44 @@ GenericTypeOrExtensionWhereOrBodyPortion::computeSelfDC(
   while (i != 0) {
     Optional<NullablePtr<DeclContext>> maybeSelfDC =
         history[--i]->computeSelfDCForParent();
-    if (maybeSelfDC)
-      return *maybeSelfDC;
+    if (maybeSelfDC) {
+      // If we've found a selfDC, we'll definitely be returning something.
+      // However, we may have captured 'self' somewhere down the tree, so we
+      // can't return outright without checking the nested scopes.
+      NullablePtr<DeclContext> nestedCapturedSelfDC =
+          checkNestedScopesForSelfCapture(history, i);
+      return nestedCapturedSelfDC ?: *maybeSelfDC;
+    }
   }
   return nullptr;
 }
 
+NullablePtr<DeclContext>
+GenericTypeOrExtensionWhereOrBodyPortion::checkNestedScopesForSelfCapture(
+    ArrayRef<const ASTScopeImpl *> history, size_t start) {
+  NullablePtr<DeclContext> innerCapturedSelfDC;
+  // Start with the next scope down the tree.
+  size_t j = start;
+  while (j != 0) {
+    Optional<NullablePtr<DeclContext>> maybeCapturedSelfDC =
+        history[--j]->computeSelfDCForParent();
```

Comment:
I posted a comment with some thoughts in the thread on your previous review, but that got buried by these most recent changes. Copy/pasting here for visibility. (And, if you agree with this reasoning and think it’s appropriate to have a comment here explaining *why* this nested loop doesn’t cause a quadratic blowup, I’m happy to add that!)

Reply:
> Have considered this, and I think if we switch to a top-down approach here rather than a bottom up, we should be able to avoid any quadratic behavior. Currently, if we have a structure like:

> ```
struct A {
    var x: Int = 0
    func a() {
        struct B {
            func b() {
            // ...
                struct Z {
                    func z() { print(x) }
                }
            // ...
            }
        }
    }
}
```

> Then the lookup will, each time we reach a new type scope, traverse the entire history so far (beginning from the local scope of `x`, leading to a quadratic blowup. But if we instead begin looking for captured `self` contexts *at* the type scope and then drill down until we hit something that causes us to forget the `self` context, (i.e., a `GenericTypeOrExtensionScope`), then we'll only re-traverse the scope history at most one additional time total.

> You're much more familiar with this code than I am, but let me know if that sounds right to you. I'll bang out an implementation going in that direction.

Reply:
How would it work to thread a parameter through the recursion? Would that be clearer? As is, his function reminds me of `computeIsCascadingUse`, IIRC, I only resorted to a post-hoc historical inspection because the thread-through-the-recursion option seemed quite opaque. Is there a sensible description for what the extra parameter would be?

Reply:
After looking at it and thinking, I think it's fine.